File: examples/DeviceBuilder.java
/*
 * An example for implementing the device builder interface on DoubleSolenoid.
 */

import edu.wpi.first.wpilibj.DoubleSolenoid;
import edu.wpi.first.wpilibj.PneumaticsModuleType;
import com.redstorm509.stormkit.interfaces.IDeviceBuilder;

public class DeviceBuilder {
	public static class DoubleSolenoidBuilder implements IDeviceBuilder<DoubleSolenoid> {
		final int id;
		final int forwardChannel;
		final int reverseChannel;
		final PneumaticsModuleType type;

		public DoubleSolenoidBuilder(int id, int forwardChannel, int reverseChannel, PneumaticsModuleType type) {
			this.id = id;
			this.forwardChannel = forwardChannel;
			this.reverseChannel = reverseChannel;
			this.type = type;
		}

		@Override
		public DoubleSolenoid build() {
			return new DoubleSolenoid(id, type, forwardChannel, reverseChannel);
		}
	}
}

File: examples/DeviceBuilder.cpp
/*
 * An example for implementing the device builder interface on DoubleSolenoid.
 */

#include <frc/DoubleSolenoid.h>
#include <stormkit/interfaces/DeviceBuilder.hpp>

template<> class DeviceBuilder<frc::DoubleSolenoid> {
private:
    int id;
    int forward_channel;
    int reverse_channel;
    frc::PneumaticsModuleType type;
public:
	DeviceBuilder(int id, int forward_channel, int reverse_channel, frc::PneumaticsModuleType type) {
		this->id = id;
        this->forward_channel = forward_channel;
        this->reverse_channel = reverse_channel;
        this->type = type;
	}
	frc::DoubleSolenoid Build() {
		return frc::DoubleSolenoid{id, type, forward_channel, reverse_channel};
	}
};